File: sdks/python/fm/events.py
"""WebSocket STOMP event listener for Flexemarkets.

Implements STOMP 1.2 framing over WebSocket, matching the Java
EventParser / Flexemarkets.listen() behaviour.
"""

from __future__ import annotations

import json
import logging
import os
import queue
import threading
import time
from dataclasses import dataclass, field
from typing import Any

import websockets.sync.client as ws_sync

from .types import Holding, Order, Session, Version

log = logging.getLogger(__name__)

_HEARTBEAT_MS = 30_000

# How often this client actually sends one. Under the 30s it advertises, so a
# late timer still lands inside the window it promised, and well under the 55s
# at which Heroku's router closes an idle connection and records an H15
# against the app. Named so the interval can be tested against the thing it
# exists to stay under.
_HEARTBEAT_INTERVAL_SECONDS = 25.0
_HEROKU_IDLE_TIMEOUT_MS = 55_000
_INBOUND_MESSAGE_SIZE = 128 * 1024 * 1024  # 128 MB
_NULL = "\x00"


def _resolve_api_version_prefix() -> str:
    """Prefix on the /app SUBSCRIBE destination selecting fm-server's
    WS API version. Empty string -> V0 (/app/marketplaces/{id});
    "/v1" -> V1 (/app/v1/marketplaces/{id}). V1 omits the bulk
    ORDERS-UPDATE snapshot on subscribe, keeping inbound frames small.
    V1 is the default; override with FM_WS_API_VERSION=v0 if talking
    to an old fm-server that doesn't speak V1.

    NB: V1 SUBSCRIBE delivers an empty ORDERS-UPDATE; consumers that
    need the active book at startup should fetch it via REST
    (GET /api/v1/marketplaces/{id}/orders/active) and reconcile
    against incoming deltas using the seq header.
    """
    v = os.environ.get("FM_WS_API_VERSION", "v1").strip().lower()
    if v == "v0":
        return ""
    if v == "v1":
        return "/v1"
    raise ValueError(f"FM_WS_API_VERSION must be 'v0' or 'v1', got: {v}")


_API_VERSION_PREFIX = _resolve_api_version_prefix()


# ---------------------------------------------------------------------------
# STOMP frame codec
# ---------------------------------------------------------------------------

@dataclass
class StompFrame:
    command: str = ""
    headers: dict[str, str] = field(default_factory=dict)
    body: str = ""


def _encode_frame(frame: StompFrame) -> str:
    lines = [frame.command]
    for k, v in frame.headers.items():
        lines.append(f"{k}:{v}")
    lines.append("")
    lines.append(frame.body)
    return "\n".join(lines) + _NULL


def _decode_frame(raw: str) -> StompFrame:
    # Strip trailing NUL and whitespace
    raw = raw.rstrip(_NULL)

    # Split command + headers from body at the first blank line
    header_end = raw.find("\n\n")
    if header_end < 0:
        # No body
        header_section = raw
        body = ""
    else:
        header_section = raw[:header_end]
        body = raw[header_end + 2:]

    lines = header_section.split("\n")
    command = lines[0] if lines else ""
    headers: dict[str, str] = {}
    for line in lines[1:]:
        if ":" in line:
            k, v = line.split(":", 1)
            headers[k] = v

    return StompFrame(command=command, headers=headers, body=body)


# ---------------------------------------------------------------------------
# Error / transport event types put onto the queue
# ---------------------------------------------------------------------------

def _is_auth_refusal(failure: BaseException) -> bool:
    """Whether the handshake was refused for who we are, not for what the
    network did.

    ``websockets`` raises :class:`InvalidStatus` carrying the response, so the
    status is read off that rather than out of a message the library owns. 403
    counts too: a token that is valid but not permitted here will not become
    permitted by being presented again.
    """
    cause: BaseException | None = failure
    seen: set[int] = set()
    while cause is not None and id(cause) not in seen:
        seen.add(id(cause))
        response = getattr(cause, "response", None)
        status = getattr(response, "status_code", None)
        if status in (401, 403):
            return True
        cause = cause.__cause__ or cause.__context__
    return False


@dataclass
class StreamDropped:
    exception: BaseException


@dataclass
class FrameUnreadable:
    command: str
    headers: dict[str, str]
    body: str
    exception: BaseException


@dataclass
class Reconnected:
    """The stream is back.

    Put on the queue once the transport has re-established itself after a
    :class:`StreamDropped`. A consumer's state is stale until it reseeds:
    whatever happened while the socket was down was not delivered.

    Carries the marketplace because one client can hold several
    subscriptions, and "we are back" does not say which desk to reseed.
    """

    marketplace_id: int


# Sentinel for "no seq header" — see OrdersUpdate.seq docstring.
NO_SEQ = -1


@dataclass
class OrdersUpdate:
    """One ORDERS-UPDATE delta from the WS stream, carrying the parsed
    order list plus the per-marketplace ``seq`` header fm-server stamps
    each frame with (commit c6eea6eca).

    Consumers reconcile this against a REST :class:`Snapshot`: apply
    deltas whose :attr:`seq` is greater than the snapshot's
    ``as_of_seq`` and skip those whose seq is less than or equal.

    :attr:`seq` is :data:`NO_SEQ` when the server didn't stamp the
    header (older fm-server).
    """

    orders: list[Order]
    seq: int


# ---------------------------------------------------------------------------
# Event parsing — mirrors Java EventParser.getPayloadType()
# ---------------------------------------------------------------------------

_MESSAGE_TYPE = "message-type"
_SEQ_HEADER   = "seq"


def _parse_event(frame: StompFrame) -> object | None:
    msg_type = frame.headers.get(_MESSAGE_TYPE)
    if msg_type is None:
        log.warning("Unresolved payload type, headers: %s", frame.headers)
        return None

    try:
        data = json.loads(frame.body) if frame.body else None
    except json.JSONDecodeError:
        log.warning("Failed to decode JSON body for %s", msg_type)
        return None

    match msg_type:
        case "VERSION":
            return _parse_version(data)
        case "SESSION-LIST":
            return _parse_session_list(data)
        case "SESSION-UPDATE":
            return _parse_session(data)
        case "HOLDING-UPDATE":
            return _parse_holding(data)
        case "ORDERS-UPDATE":
            return OrdersUpdate(
                orders=_parse_orders(data),
                seq=_parse_seq(frame.headers.get(_SEQ_HEADER)),
            )
        case _:
            log.warning("Unknown message-type: %s", msg_type)
            return None


def _parse_seq(value: str | None) -> int:
    """Parse the per-frame ``seq`` STOMP header; falls back to
    :data:`NO_SEQ` on absent or malformed values.
    """
    if value is None:
        return NO_SEQ
    try:
        return int(value)
    except ValueError:
        return NO_SEQ


def _parse_version(data: Any) -> Version:
    if isinstance(data, dict):
        return Version(version=data.get("version", 0))
    return Version(version=int(data) if data else 0)


def _parse_session(data: dict[str, Any]) -> Session:
    from .client import _parse_session as _client_parse_session
    return _client_parse_session(data)


def _parse_session_list(data: Any) -> list[Session]:
    if isinstance(data, list):
        return [_parse_session(s) for s in data]
    return []


def _parse_holding(data: dict[str, Any]) -> Holding:
    from .client import _parse_holding as _client_parse_holding
    return _client_parse_holding(data)


def _parse_order(data: dict[str, Any]) -> Order:
    from .client import _parse_order as _client_parse_order
    return _client_parse_order(data)


def _parse_orders(data: Any) -> list[Order]:
    if isinstance(data, list):
        return [_parse_order(o) for o in data]
    return []


# ---------------------------------------------------------------------------
# EventListener — background thread that manages WS + STOMP
# ---------------------------------------------------------------------------

class EventListener:
    """Manages a STOMP-over-WebSocket connection and pushes typed events
    into a :class:`queue.Queue`.

    This mirrors the Java ``Flexemarkets.listen()`` + ``EventParser``
    combination.
    """

    def __init__(
        self,
        ws_url: str,
        bearer_token: str,
        marketplace_id: int,
        event_queue: queue.Queue[object],
        client_description: str = "Unspecified client",
    ):
        self._ws_url = ws_url
        self._bearer_token = bearer_token
        self._marketplace_id = marketplace_id
        self._queue = event_queue
        self._client_description = client_description

        self._ws: ws_sync.ClientConnection | None = None
        self._thread: threading.Thread | None = None
        self._closed = False
        self._subscription_counter = 0
        self._reconnecting = threading.Lock()

        # Sends the heartbeats this client promises.
        #
        # IT PROMISED THEM AND NEVER SENT ONE. The CONNECT frame has always
        # carried heart-beat:30000,30000 -- "I will send every 30s, send me
        # one every 30s" -- and nothing here wrote to the socket again except
        # SUBSCRIBE. The websockets library's own 20s ping kept the connection
        # off Heroku's idle timer, so nothing broke visibly; but the
        # advertisement was still a claim this client could not back, and a
        # ping is not a STOMP heartbeat. Java fixed this first; Python and
        # TypeScript advertised the same thing and were left behind.
        #
        # A daemon thread, so it can never hold the interpreter open.
        self._heartbeat_stop = threading.Event()
        self._heartbeat_thread: threading.Thread | None = None

    # -- public API --------------------------------------------------------

    def start(self) -> None:
        """Connect and begin receiving events in a background thread."""
        self._ws = self._connect()
        self._stomp_connect()
        self._subscribe()
        self._start_heartbeats()

        self._thread = threading.Thread(
            target=self._receive_loop,
            name="fm-ws-events",
            daemon=True,
        )
        self._thread.start()

    def _on_stream_dropped(self, exc: BaseException) -> None:
        """React to a dropped stream by restoring it, off the receive thread.

        The receive loop is dead by the time this runs and reconnect() sleeps
        between attempts, so the work goes to its own thread. On success a
        :class:`Reconnected` goes onto the queue so consumers know their state
        needs reseeding -- matching Java, where the subscription has always
        restored itself rather than leaving the caller to notice.

        The lock is what makes a burst of errors from a dying socket one
        outage instead of five reconnects; Java uses compareAndSet on a flag
        for the same reason.
        """
        if self._closed:
            return
        self._queue.put(StreamDropped(exception=exc))
        if not self._reconnecting.acquire(blocking=False):
            return

        def _restore() -> None:
            try:
                self.reconnect()
                if not self._closed:
                    self._queue.put(Reconnected(marketplace_id=self._marketplace_id))
            finally:
                self._reconnecting.release()

        threading.Thread(target=_restore, name="fm-ws-reconnect", daemon=True).start()

    def reconnect(self) -> None:
        """Reconnect after a transport error.  Blocks until connected."""
        while not self._closed:
            try:
                self._disconnect()
                self._ws = self._connect()
                self._stomp_connect()
                self._subscribe()
                self._start_heartbeats()

                self._thread = threading.Thread(
                    target=self._receive_loop,
                    name="fm-ws-events",
                    daemon=True,
                )
                self._thread.start()
                return
            except Exception as exc:
                # A REJECTED TOKEN IS NOT A BLIP, and retrying it is not
                # patience -- it is a client hammering a server that has
                # already given its final answer. Measured 2026-09-03: one NAT
                # gateway produced 11,918 handshake 401s in two hours, about
                # 100 a minute, against 360 successful connections from every
                # other client combined. Nothing told the operator their token
                # had expired; the loop simply ran.
                #
                # So an auth refusal ends the stream and says why, and every
                # other failure keeps the existing retry: a server restart or a
                # network drop IS a blip, and recovering from it is what this
                # loop is for.
                if _is_auth_refusal(exc):
                    self._closed = True
                    self._queue.put(StreamDropped(exception=PermissionError(
                        "WebSocket refused: the token was rejected. It has expired "
                        "or was signed for another server; reconnecting cannot fix it."
                    )))
                    return
                time.sleep(2)

    def close(self) -> None:
        """Disconnect and stop the background thread."""
        if self._closed:
            return
        self._closed = True
        self._stop_heartbeats()
        self._disconnect()

    def _start_heartbeats(self) -> None:
        """Send a bare EOL every :data:`_HEARTBEAT_INTERVAL_SECONDS`.

        A newline on its own is what STOMP 1.2 defines a heartbeat to be, so
        the server reads it as one and nothing else parses it as a frame.
        """
        self._stop_heartbeats()
        self._heartbeat_stop = threading.Event()

        def _beat() -> None:
            while not self._heartbeat_stop.wait(_HEARTBEAT_INTERVAL_SECONDS):
                ws = self._ws
                if self._closed or ws is None:
                    return
                try:
                    ws.send("\n")
                except Exception:  # a dead socket is the receive loop's problem
                    return

        self._heartbeat_thread = threading.Thread(
            target=_beat, name="fm-sdk-stomp-heartbeat", daemon=True
        )
        self._heartbeat_thread.start()

    def _stop_heartbeats(self) -> None:
        self._heartbeat_stop.set()
        self._heartbeat_thread = None

    # -- WebSocket connection ----------------------------------------------

    def _connect(self) -> ws_sync.ClientConnection:
        return ws_sync.connect(
            self._ws_url,
            additional_headers={
                "Authorization": self._bearer_token,
            },
            subprotocols=["v12.stomp", "v11.stomp", "v10.stomp"],
            max_size=_INBOUND_MESSAGE_SIZE,
        )

    def _disconnect(self) -> None:
        if self._ws is not None:
            try:
                self._ws.close()
            except Exception:
                pass
            self._ws = None

    # -- STOMP framing -----------------------------------------------------

    def _next_id(self) -> str:
        self._subscription_counter += 1
        return f"sub-{self._subscription_counter}"

    def _stomp_connect(self) -> None:
        frame = StompFrame(
            command="CONNECT",
            headers={
                "accept-version": "1.2,1.1,1.0",
                "heart-beat": f"{_HEARTBEAT_MS},{_HEARTBEAT_MS}",
                "agent-description": self._client_description,
                "marketplace-id": str(self._marketplace_id),
            },
        )
        self._ws.send(_encode_frame(frame))  # type: ignore[union-attr]

        # Wait for CONNECTED frame
        raw = self._ws.recv()  # type: ignore[union-attr]
        reply = _decode_frame(raw)
        if reply.command not in ("CONNECTED", ""):
            log.debug("STOMP CONNECT reply: %s", reply.command)

    def _subscribe(self) -> None:
        # fm-server publishes broadcasts on the V0 destination paths
        # (/topic/marketplaces/{id}, /user/queue/marketplaces/{id}) for
        # both V0 and V1 clients — only the @SubscribeMapping gating
        # the initial snapshot lives at the /v1 prefix. So pub/sub
        # subscriptions stay on V0 paths regardless of the chosen
        # api-version; only the /app destination flips. Mirrors
        # fm-ui's web-socket.service.ts and the Java/TS SDKs.
        resource = f"/marketplaces/{self._marketplace_id}"
        for dest in (
            f"/user/queue{resource}",
            f"/topic{resource}",
            f"/app{_API_VERSION_PREFIX}{resource}",
        ):
            frame = StompFrame(
                command="SUBSCRIBE",
                headers={
                    "id": self._next_id(),
                    "destination": dest,
                },
            )
            self._ws.send(_encode_frame(frame))  # type: ignore[union-attr]

    # -- receive loop (background thread) ----------------------------------

    def _receive_loop(self) -> None:
        try:
            while not self._closed:
                try:
                    raw = self._ws.recv(timeout=_HEARTBEAT_MS / 1000 * 2)  # type: ignore[union-attr]
                except TimeoutError:
                    continue

                if not raw or raw == "\n":
                    # STOMP heartbeat
                    continue

                frame = _decode_frame(raw)

                if frame.command == "MESSAGE":
                    event = _parse_event(frame)
                    if event is not None:
                        self._queue.put(event)
                elif frame.command == "ERROR":
                    self._queue.put(
                        FrameUnreadable(
                            command=frame.command,
                            headers=frame.headers,
                            body=frame.body,
                            exception=RuntimeError(frame.headers.get("message", "STOMP ERROR")),
                        )
                    )
                # Ignore RECEIPT, CONNECTED, heartbeats, etc.

        except Exception as exc:
            self._on_stream_dropped(exc)

File: sdks/java/fm-sdk/src/main/java/fm/internal/Events.java
package fm.internal;

import fm.event.FrameUnreadable;
import fm.event.OrdersUpdate;
import fm.event.StreamReconnected;
import fm.event.StreamDropped;
import fm.event.Version;
import fm.role.Streaming;
import fm.model.Holding;
import fm.model.Order;
import fm.model.Session;
import fm.Flexemarkets;
import fm.Desk;
import fm.Snapshot;
import fm.Subscription;
import fm.error.ApiException;
import fm.error.AuthenticationException;
import fm.error.FlexemarketsException;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.WebSocket;
import java.net.http.WebSocketHandshakeException;
import java.nio.ByteBuffer;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import tools.jackson.core.type.TypeReference;
import tools.jackson.databind.ObjectMapper;


/**
 * The STOMP client underneath {@link fm.Streaming}: one WebSocket
 * subscription, delivering onto a caller's queue.
 *
 * <p>Internal because it is the implementation. A caller reaches this through
 * {@code listen}, {@code subscribe} or {@code desk}, and holds the
 * {@link Subscription} rather than this type.
 */
public class Events implements Subscription {

    /**
     * Prefix on the {@code /app} SUBSCRIBE destination selecting fm-server's
     * WS API version. Empty string → V0 ({@code /app/marketplaces/{id}});
     * {@code "/v1"} → V1 ({@code /app/v1/marketplaces/{id}}). V1 omits the
     * bulk ORDERS-UPDATE snapshot on subscribe, keeping inbound frames
     * small. V1 is the default; override with
     * {@code -Dfm.net.ws.api-version=v0} if talking to an old fm-server
     * that doesn't speak V1.
     *
     * <p>NB: V1 SUBSCRIBE delivers an empty ORDERS-UPDATE; consumers that
     * need the active book at startup should fetch it via REST
     * ({@code GET /api/v1/marketplaces/{id}/orders/active}) and reconcile
     * against incoming deltas using the {@code seq} header.
     */
    private static final String API_VERSION_PREFIX = _resolveApiVersionPrefix();

    private static String _resolveApiVersionPrefix() {
        String version = System.getProperty("fm.net.ws.api-version", "v1").trim();
        if ("v0".equalsIgnoreCase(version)) return "";
        if ("v1".equalsIgnoreCase(version)) return "/v1";
        throw new IllegalArgumentException(
                "fm.net.ws.api-version must be 'v0' or 'v1', got: " + version);
    }

    private static final String MESSAGE_TYPE = "message-type";
    private static final String SEQ          = "seq";

    private static final String MESSAGE_TYPE_VERSION        = "VERSION";
    private static final String MESSAGE_TYPE_SESSION_LIST   = "SESSION-LIST";
    private static final String MESSAGE_TYPE_SESSION_UPDATE = "SESSION-UPDATE";
    private static final String MESSAGE_TYPE_HOLDING_UPDATE = "HOLDING-UPDATE";
    private static final String MESSAGE_TYPE_ORDERS_UPDATE  = "ORDERS-UPDATE";

    private static final TypeReference<Version>    VERSION_TYPE  = new TypeReference<>() {};
    private static final TypeReference<Session[]>  SESSIONS_TYPE = new TypeReference<>() {};
    private static final TypeReference<Session>    SESSION_TYPE  = new TypeReference<>() {};
    private static final TypeReference<Holding>    HOLDING_TYPE  = new TypeReference<>() {};
    private static final TypeReference<Order[]>    ORDERS_TYPE   = new TypeReference<>() {};

    private final String _wsUrl;
    private final String _bearerToken;
    private final long _marketplaceId;
    private final String _clientDescription;
    private final ObjectMapper _mapper;
    private final BlockingQueue<Object> _queue;

    /**
     * How often a heartbeat goes out, under the 30s this client ADVERTISES in
     * CONNECT so it is never late, and well under the 55 seconds after which
     * Heroku's router closes an idle connection and records an H15.
     */
    static final long HEARTBEAT_INTERVAL_SECONDS = 25;

    /** What this client advertises in CONNECT, and must therefore keep to. */
    static final long ADVERTISED_HEARTBEAT_MS = 30_000;

    /**
     * After this much silence in either direction, Heroku's router closes the
     * connection and records an H15 against the app. Here so the interval
     * above can be tested against the thing it exists to stay under.
     */
    static final long HEROKU_IDLE_TIMEOUT_MS = 55_000;

    private volatile WebSocket _webSocket;
    private volatile boolean _closed;

    /**
     * Sends the heartbeats this client promises.
     *
     * <p>IT PROMISED THEM AND NEVER SENT ONE. The CONNECT frame has always
     * carried `heart-beat:30000,30000` -- "I will send every 30s, send me one
     * every 30s" -- and nothing in this class ever wrote to the socket again
     * except SUBSCRIBE. The server's own heartbeats covered for it, so nothing
     * broke; but a robot that idles is relying on the far end to keep its
     * socket alive, and the advertisement was a claim this client could not
     * back. A daemon thread, so it can never hold a JVM open.
     */
    private final ScheduledExecutorService _heartbeats =
            Executors.newSingleThreadScheduledExecutor(r -> {
                var thread = new Thread(r, "fm-sdk-stomp-heartbeat");
                thread.setDaemon(true);
                return thread;
            });
    private volatile ScheduledFuture<?> _heartbeat;

    /** One reconnect at a time: onClose and onError can both fire for one drop. */
    private final java.util.concurrent.atomic.AtomicBoolean _reconnecting =
            new java.util.concurrent.atomic.AtomicBoolean();
    private final AtomicInteger _subscriptionId = new AtomicInteger(0);

    Events(String wsUrl, String bearerToken, long marketplaceId, String clientDescription,
           ObjectMapper mapper, BlockingQueue<Object> queue) {
        this._wsUrl = wsUrl;
        this._bearerToken = bearerToken;
        this._marketplaceId = marketplaceId;
        this._clientDescription = clientDescription;
        this._mapper = mapper;
        this._queue = queue;
    }

    void connect() {
        try {
            var connectedLatch = new CountDownLatch(1);
            var listener = new StompListener(connectedLatch);

            // NORMAL rather than the JDK's default of NEVER, for the same reason
            // the REST client uses it: an edge that upgrades plain HTTP to HTTPS
            // answers the handshake with a 301 the client must follow.
            this._webSocket = HttpClient.newBuilder()
                .followRedirects(HttpClient.Redirect.NORMAL)
                .build()
                .newWebSocketBuilder()
                .header("Authorization", _bearerToken)
                .subprotocols("v12.stomp", "v11.stomp", "v10.stomp")
                .buildAsync(URI.create(_wsUrl), listener)
                .join();

            _sendStompConnect();

            if (!connectedLatch.await(10, TimeUnit.SECONDS)) {
                throw new ApiException("STOMP CONNECTED frame not received within timeout");
            }

            // fm-server publishes broadcasts on the V0 destination paths
            // (/topic/marketplaces/{id}, /user/queue/marketplaces/{id})
            // for both V0 and V1 clients — only the @SubscribeMapping
            // gating the initial snapshot lives at the /v1 prefix. So
            // pub/sub subscriptions stay on V0 paths regardless of the
            // chosen api-version; only the /app destination flips.
            // Mirrors fm-ui's web-socket.service.ts and fm-robots'
            // EventParser pattern.
            _subscribe("/user/queue/marketplaces/" + _marketplaceId);
            _subscribe("/topic/marketplaces/" + _marketplaceId);
            _subscribe("/app" + API_VERSION_PREFIX + "/marketplaces/" + _marketplaceId);

            _startHeartbeats();
        } catch (FlexemarketsException e) {
            throw e;
        } catch (Exception e) {
            throw new ApiException("WebSocket connection failed", e);
        }
    }

    /**
     * Re-establish the stream, retrying until it succeeds or this is closed.
     *
     * <p>Driven from {@link #reconnectInBackground()} rather than by a
     * consumer: whether the socket is up is this class's business, and a
     * consumer that had to run the retry loop would be doing the transport's
     * job on its own thread.
     *
     * <p>Package-private, not private, because {@link Flexemarkets#reconnect()}
     * still lets a caller force one deliberately. That is a different thing
     * from recovering a drop, and only the recovery moved in here.
     */
    void reconnect() throws InterruptedException {
        while (!_closed) {
            try {
                _closeWebSocket();
                connect();
                return;
            } catch (Exception e) {
                // A REJECTED TOKEN IS NOT A BLIP, and retrying it is not
                // patience -- it is a client hammering a server that has
                // already given its final answer. Measured 2026-09-03: one NAT
                // gateway produced 11,918 handshake 401s in two hours, ~100 a
                // minute, against 360 successful connections from every other
                // client combined. Nothing told the operator their token had
                // expired; the loop simply ran.
                //
                // So an auth refusal ends the stream and says why, and every
                // other failure keeps the existing retry: a server restart or
                // a network drop IS a blip and recovering from it is the whole
                // point of this loop.
                if (_isAuthRefusal(e)) {
                    _closed = true;
                    _queue.offer(new StreamDropped(new AuthenticationException(
                        "WebSocket refused: the token was rejected. It has expired or was "
                        + "signed for another server; reconnecting cannot fix it.", e)));
                    return;
                }
                TimeUnit.SECONDS.sleep(2);
            }
        }
    }

    /**
     * Whether the handshake was refused for who we are rather than for what
     * the network did.
     *
     * <p>The status is read off the handshake response rather than a message,
     * because the text is the JDK's and could change under us. 403 counts too:
     * a token that is valid but not permitted here will not become permitted
     * by being presented again.
     */
    private static boolean _isAuthRefusal(Throwable failure) {
        for (var cause = failure; cause != null; cause = cause.getCause()) {
            if (cause instanceof WebSocketHandshakeException handshake) {
                var status = handshake.getResponse().statusCode();
                return status == 401 || status == 403;
            }
            if (cause == cause.getCause()) break;
        }
        return false;
    }

    /**
     * React to a dropped stream by restoring it, off the callback thread.
     *
     * <p>The WebSocket listener must not block, and reconnect() sleeps between
     * attempts. On success a {@link StreamReconnected} goes onto the queue so
     * consumers know their state needs reseeding.
     *
     * <p>Package-private for the same reason {@link #reconnect()} is: the
     * recovery path is otherwise reachable only through a live socket, since
     * the listener that calls it is built inside {@link #connect()}. Driving
     * it directly is what lets StreamReconnectTest cover the behaviour Python
     * and TypeScript have covered all along.
     */
    void reconnectInBackground() {
        if (_closed || !_reconnecting.compareAndSet(false, true)) {
            return;
        }
        Thread.startVirtualThread(() -> {
            try {
                reconnect();
                if (!_closed) {
                    _queue.offer(new StreamReconnected(_marketplaceId));
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } finally {
                _reconnecting.set(false);
            }
        });
    }

    @Override
    public void close() {
        if (_closed) return;
        _closed = true;
        _closeWebSocket();
        _heartbeats.shutdownNow();
    }

    private void _closeWebSocket() {
        _stopHeartbeats();
        if (_webSocket != null) {
            try {
                _webSocket.sendClose(WebSocket.NORMAL_CLOSURE, "").join();
            } catch (Exception ignored) {}
        }
    }

    /**
     * Begin sending the STOMP heartbeat: a bare end-of-line, which is what the
     * protocol defines a heartbeat to be.
     *
     * <p>Started only after CONNECTED. Before that there is no STOMP session
     * to heartbeat on, and a stray newline into a socket mid-handshake is a
     * frame the server has no reason to expect.
     *
     * <p>Failures are swallowed on purpose. A heartbeat that cannot be written
     * means the socket is gone, and the listener's onClose/onError is what
     * reconnects -- raising from a scheduled task would kill the schedule and
     * report a fault the transport is already handling.
     */
    private void _startHeartbeats() {
        _stopHeartbeats();
        _heartbeat = _heartbeats.scheduleAtFixedRate(() -> {
            var socket = _webSocket;
            if (_closed || socket == null) return;
            try {
                socket.sendText("\n", true);
            } catch (Exception ignored) {}
        }, HEARTBEAT_INTERVAL_SECONDS, HEARTBEAT_INTERVAL_SECONDS, TimeUnit.SECONDS);
    }

    private void _stopHeartbeats() {
        var running = _heartbeat;
        if (running != null) {
            running.cancel(false);
            _heartbeat = null;
        }
    }

    // --- STOMP frame encoding/decoding ---

    private void _sendStompConnect() {
        var frame = _stompFrame("CONNECT",
            List.of(
                "accept-version:1.2",
                "heart-beat:" + ADVERTISED_HEARTBEAT_MS + "," + ADVERTISED_HEARTBEAT_MS,
                "agent-description:" + _clientDescription,
                "marketplace-id:" + _marketplaceId
            ),
            null);
        _webSocket.sendText(frame, true);
    }

    private void _subscribe(String destination) {
        var id = "sub-" + _subscriptionId.getAndIncrement();
        var frame = _stompFrame("SUBSCRIBE",
            List.of("id:" + id, "destination:" + destination),
            null);
        _webSocket.sendText(frame, true);
    }

    private static String _stompFrame(String command, List<String> headers, String body) {
        var sb = new StringBuilder();
        sb.append(command).append('\n');
        for (var header : headers) {
            sb.append(header).append('\n');
        }
        sb.append('\n');
        if (body != null) {
            sb.append(body);
        }
        sb.append('\0');
        return sb.toString();
    }

    private void _dispatchStompMessage(String frame) {
        var lines = frame.split("\n", -1);
        if (lines.length < 2) return;

        var command = lines[0].trim();

        if ("CONNECTED".equals(command)) {
            return; // handled by latch in listener
        }

        if ("ERROR".equals(command)) {
            _queue.offer(new FrameUnreadable("STOMP ERROR: " + frame, null));
            return;
        }

        if (!"MESSAGE".equals(command)) {
            return;
        }

        String messageType = null;
        long seq = Snapshot.NO_SEQ;
        int bodyStart = -1;

        for (int i = 1; i < lines.length; i++) {
            var line = lines[i];
            if (line.isEmpty()) {
                bodyStart = i + 1;
                break;
            }
            if (line.startsWith(MESSAGE_TYPE + ":")) {
                messageType = line.substring(MESSAGE_TYPE.length() + 1).trim();
            } else if (line.startsWith(SEQ + ":")) {
                // Per-marketplace ORDERS-UPDATE sequence number stamped
                // by fm-server (commit c6eea6eca). Used by Desk /
                // Phase 2a snapshot reconciliation. Parse defensively —
                // a malformed value just falls back to NO_SEQ.
                try {
                    seq = Long.parseLong(line.substring(SEQ.length() + 1).trim());
                } catch (NumberFormatException ignored) { /* leave as NO_SEQ */ }
            }
        }

        if (messageType == null || bodyStart < 0) return;

        var bodyBuilder = new StringBuilder();
        for (int i = bodyStart; i < lines.length; i++) {
            if (i > bodyStart) bodyBuilder.append('\n');
            bodyBuilder.append(lines[i]);
        }
        var body = bodyBuilder.toString();
        // strip trailing null byte
        if (body.endsWith("\0")) {
            body = body.substring(0, body.length() - 1);
        }

        try {
            // ORDERS-UPDATE gets wrapped so the seq header reaches
            // consumers; everything else is pushed as the parsed
            // payload directly. This is a breaking change for callers
            // that did `case Order[] orders` — they need to switch to
            // `case OrdersUpdate update` and read update.orders().
            Object event = switch (messageType) {
                case MESSAGE_TYPE_VERSION        -> _mapper.readValue(body, VERSION_TYPE);
                case MESSAGE_TYPE_SESSION_LIST   -> _mapper.readValue(body, SESSIONS_TYPE);
                case MESSAGE_TYPE_SESSION_UPDATE -> _mapper.readValue(body, SESSION_TYPE);
                case MESSAGE_TYPE_HOLDING_UPDATE -> _mapper.readValue(body, HOLDING_TYPE);
                case MESSAGE_TYPE_ORDERS_UPDATE  -> new OrdersUpdate(_mapper.readValue(body, ORDERS_TYPE), seq);
                default -> null;
            };
            if (event != null) {
                _queue.offer(event);
            }
        } catch (Exception e) {
            _queue.offer(new FrameUnreadable("Failed to parse STOMP message: " + messageType, e));
        }
    }

    // --- WebSocket.Listener ---

    private class StompListener implements WebSocket.Listener {
        private final CountDownLatch _connectedLatch;
        private final StringBuilder _buffer = new StringBuilder();

        StompListener(CountDownLatch connectedLatch) {
            this._connectedLatch = connectedLatch;
        }

        @Override
        public void onOpen(WebSocket webSocket) {
            webSocket.request(1);
        }

        @Override
        public CompletionStage<?> onText(WebSocket webSocket, CharSequence data, boolean last) {
            _buffer.append(data);
            if (last) {
                var frame = _buffer.toString();
                _buffer.setLength(0);

                if (frame.startsWith("CONNECTED")) {
                    _connectedLatch.countDown();
                }

                // Dispatch in a virtual thread to avoid blocking the WS receive thread
                Thread.startVirtualThread(() -> _dispatchStompMessage(frame));
            }
            webSocket.request(1);
            return CompletableFuture.completedFuture(null);
        }

        @Override
        public CompletionStage<?> onPing(WebSocket webSocket, ByteBuffer message) {
            webSocket.sendPong(message);
            webSocket.request(1);
            return CompletableFuture.completedFuture(null);
        }

        @Override
        public CompletionStage<?> onClose(WebSocket webSocket, int statusCode, String reason) {
            if (!_closed) {
                _queue.offer(new StreamDropped(
                    new Exception("WebSocket closed: %d %s".formatted(statusCode, reason))));
                reconnectInBackground();
            }
            return CompletableFuture.completedFuture(null);
        }

        @Override
        public void onError(WebSocket webSocket, Throwable error) {
            _queue.offer(new StreamDropped(error));
            reconnectInBackground();
        }
    }
}
